core/thread/inc/ThreadPool.h

```diff
@@ -0,0 +1,106 @@
+#ifndef THREADPOOL_H_
+#define THREADPOOL_H_
+
+#include "tbb/tbb.h"
+#include "TPool.h"
+
+class ThreadPool: public TPool<ThreadPool> {
+public:
+   explicit ThreadPool(){
+      this->initTBB->initialize();
+   }
+
+   explicit ThreadPool(size_t nThreads){
+      this->initTBB->initialize(nThreads);
+   }
+
+   ~ThreadPool() {
+      this->initTBB->terminate();
+   }
+
+
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   /// \cond
+   template<class F, class INTEGER> auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector <decltype(func(args.front()))>;
+   template<class F, class T> auto Map(F func, std::vector<T> &args) -> std::vector<decltype(func(args.front()))>;
+   // / \endcond
+   using TPool<ThreadPool>::Map;
+
+   template<class T, class R> auto Reduce(const std::vector<T> &objs, R redfunc) -> decltype(redfunc(objs.front(), objs.front()));
+   using TPool<ThreadPool>::Reduce;
+
+private:
+   tbb::task_scheduler_init *initTBB = new tbb::task_scheduler_init(tbb::task_scheduler_init::deferred);
+
+};
+
+/************ TEMPLATE METHODS IMPLEMENTATION ******************/
+
+//////////////////////////////////////////////////////////////////////////
+/// Execute func (with no arguments) nTimes in parallel.
+/// A vector containg executions' results is returned.
+/// Functions that take more than zero arguments can be executed (with
+/// fixed arguments) by wrapping them in a lambda or with std::bind.
+template<class F>
+auto ThreadPool::Map(F func, unsigned nTimes) -> std::vector<decltype(func())>
+{
+   using retType = decltype(func());
+   std::vector<retType> reslist(nTimes);
+
+   tbb::parallel_for(0U, nTimes, [&](unsigned int i) {
```

Comment:
See [this comment](https://github.com/root-mirror/root/pull/148/files/fcb4aadd6967e1d7ea831aa4159d36a485722a85#r58717117)